packages/SwingSet/src/vats/comms/clist-inbound.js

```diff
@@ -0,0 +1,108 @@
+
+  function getLocalForRemote(remoteID, rref) {
+    const remote = getRemote(state, remoteID);
+    const lref = remote.fromRemote.get(rref);
+    assert(lref, `${rref} must already be in ${rname(remote)}`);
+    return lref;
+  }
+
+  function addLocalObjectForRemote(remote, roid) {
+    // The index must be allocated by them. If we allocated it, it should
+    // have been in our table already, and the fact that it isn't means
+    // they're reaching for something we haven't given them.
+    assert(
+      !parseRemoteSlot(roid).allocatedByRecipient,
+      `I don't remember giving ${roid} to ${rname(remote)}`,
+    );
+
+    // So this must be a new import. Allocate a new vat object for it, which
+    // will be the local machine's proxy for use by all other local vats, as
+    // well as third party machines.
+    const vatoid = makeVatSlot('object', true, state.nextObjectIndex);
+    state.nextObjectIndex += 1;
+
+    // remember who owns this object, to route messages later
+    state.objectTable.set(vatoid, remote.remoteID);
+
+    // they sent us ro-NN
+    remote.fromRemote.set(roid, vatoid);
+    // when we send it back, we'll send ro+NN
+    remote.toRemote.set(vatoid, flipRemoteSlot(roid));
+  }
+
+  function addLocalPromiseForRemote(remote, rpid) {
+    assert(
+      !parseRemoteSlot(rpid).allocatedByRecipient,
+      `I don't remember giving ${rpid} to ${rname(remote)}`,
+    );
+    // allocate a new p+NN, remember them as the decider, add to clist
+    const vpid = allocateUnresolvedPromise();
+    changeDeciderToRemote(vpid, remote.remoteID);
+    remote.fromRemote.set(rpid, vpid);
+    remote.toRemote.set(vpid, flipRemoteSlot(rpid));
+  }
+
+  function provideLocalForRemote(remoteID, rref) {
+    // We're receiving a slot from a remote system. If they've sent it to us
+    // previously, or if we're the ones who sent it to them earlier, it will be
+    // in the inbound table already.
+    const remote = getRemote(state, remoteID);
+    if (!remote.fromRemote.has(rref)) {
+      const { type } = parseRemoteSlot(rref);
+      if (type === 'object') {
+        addLocalObjectForRemote(remote, rref);
+      } else if (type === 'promise') {
+        addLocalPromiseForRemote(remote, rref);
+      } else {
+        throw Error(`cannot accept type ${type} from remote`);
+      }
+    }
+    return remote.fromRemote.get(rref);
+  }
+
+  function provideLocalForRemoteResult(remoteID, result) {
+    insistRemoteType('promise', result);
+    const vpid = provideLocalForRemote(remoteID, result);
+    // this asserts they had control over vpid, and that it wasn't already
+    // resolved. TODO: reject somehow rather than crash weirdly, we can't
+    // keep them from trying either
+    insistPromiseIsUnresolved(vpid);
+    changeDeciderFromRemoteToComms(vpid, remoteID);
+    subscribeRemoteToPromise(vpid, remoteID); // auto-subscribe sender
+    return vpid;
+  }
+
+  return harden({
+    provideLocalForRemote,
```

Comment:
The following is too dramatic a change to make now, but I should put it on your radar for future code you write:

This is @dtribble's pet peeve, and something we have discussed on eng-quality.  The preference is defining only closed-over state and helper functions at the top of a function. 
 When possible, define the returned object with concise methods (rather than indirection via a separately-defined function), like:

```js
  return harden({
    provideLocalForRemote(remoteID, rref) {
      // ...
    },
    getLocalForRemote(remoteID, rref) {
      // ...
     },
     // ...
  });
}
```

Reply:
I hear you, and I certainly like having a visible split between internal helper functions and exported API functions.

Is there a pattern (maybe some typescript thing) to let someone see the full set of what's being returned? Those function definitions are 4-20 lines each, so the full set doesn't fit on a single page, so that define-as-we-return pattern means I can't see the full list anywhere.

Reply:
You can define the return type of the function.  Or you could use a folding editor and collapse the definitions you're not interested in.